src/agent/coverage/src/elf.rs

```diff
@@ -0,0 +1,280 @@
+            // If given an index, return the denoted section if it exists and contains the VA.
+            let section = self
+                .elf
+                .section_headers
+                .get(index)
+                .ok_or_else(|| format_err!("section index out of bounds: {}", index))?;
+
+            if !section.vm_range().contains(&va) {
+                anyhow::bail!("VA not in section range: {:x}", va);
+            }
+
+            section
+        } else {
+            // If not given an index, try to find a containing section.
+            self.elf
+                .section_headers
+                .iter()
+                .find(|s| s.vm_range().contains(&va))
+                .ok_or_else(|| format_err!("VA not contained in any section: {:x}", va))?
+        };
+
+        Ok(section)
+    }
+}
+
+pub struct ElfSancovBasicBlockProvider<'d, 'e> {
+    ctx: ElfContext<'d, 'e>,
+    check_pc_table: bool,
+}
+
+impl<'d, 'e> ElfSancovBasicBlockProvider<'d, 'e> {
+    pub fn new(ctx: ElfContext<'d, 'e>) -> Self {
+        let check_pc_table = true;
+        Self {
+            check_pc_table,
+            ctx,
+        }
+    }
+
+    pub fn set_check_pc_table(&mut self, check: bool) {
+        self.check_pc_table = check;
+    }
+
+    pub fn provide(&mut self) -> Result<BTreeSet<u32>> {
+        let mut visitor = DelimiterVisitor::new(self.ctx);
+
+        for sym in self.ctx.elf.syms.iter() {
+            if let STT_NOTYPE = sym.st_type() {
+                visitor.visit_data_symbol(sym)?;
+            }
+        }
+
+        if self.check_pc_table {
+            if let Some(pcs_table) = visitor.delimiters.pcs_table(false) {
+                if let Ok(blocks) = self.provide_from_pcs_table(pcs_table) {
+                    return Ok(blocks);
+                }
+            }
+        }
+
+        if let Some(inline_table) = visitor.delimiters.inline_table(false) {
+            return self.provide_from_inline_table(inline_table);
+        }
+
+        anyhow::bail!("unable to find Sancov table")
+    }
+
+    pub fn provide_from_inline_table(
+        &mut self,
+        inline_table: SancovTable,
+    ) -> Result<BTreeSet<u32>> {
+        let mut visitor = InlineAccessVisitor::new(inline_table, self.ctx);
+
+        for sym in self.ctx.elf.syms.iter() {
+            visitor.visit_symbol(&sym)?;
+        }
+
+        Ok(visitor.scanner.offsets)
+    }
+
+    pub fn provide_from_pcs_table(&mut self, pcs_table: SancovTable) -> Result<BTreeSet<u32>> {
+        let vm_offset: u64 = pcs_table.offset.into();
+        let va = self.ctx.base + vm_offset;
+        let file_offset = self.ctx.va_to_file_offset(va, None)?;
+        let file_range = file_offset..(file_offset + pcs_table.size);
+
+        let table_data = self
+            .ctx
+            .data
+            .get(file_range)
+            .ok_or_else(|| format_err!("Sancov table data out of file range"))?;
+
+        // Assumes x86-64, `sizeof(uintptr_t) == 8`.
```

Comment:
We'll address this in a future PR by introducing a PC table parser abstraction, which we will use for both PE and ELF modules.